src/jsfeat/trainPattern.js

```diff
@@ -1,9 +1,27 @@
-// refactor this to accept an image/video?
-module.exports = function setupTrainPattern(img_u8, pattern_corners, pattern_preview, pattern_descriptors, pattern_corners, ctx, options) {
-  // exposed closure
-  return function train_pattern(newImg) {
+module.exports = function trainPattern(self) {
+  let jsfeat = require('jsfeat');
+  var img_u8;
+
+  self.addEventListener('message', function incomingMessage(request) {
+
+    if (typeof img_u8 === "undefined") img_u8 = new jsfeat.matrix_t(request.data.options.srcWidth, request.data.options.srcHeight, jsfeat.U8_t | jsfeat.C1_t);
+
+    // reply back from this worker to the main thread:
+    self.postMessage(setupTrainPattern(
+      request.data.img_u8_buffer,
+      request.data.pattern_corners,
+      request.data.pattern_preview,
+      request.data.pattern_descriptors,
+      request.data.imageData,
+      request.data.options)); // remember that these values are all cloned, so won't see later changes
+  });
+
+  function setupTrainPattern(img_u8_buffer, pattern_corners, pattern_preview, pattern_descriptors, imageData, options) {
     let detect_keypoints = require('./detectKeypoints.js');
 
+    // try overwriting buffer:
+    img_u8.buffer = img_u8_buffer;
```

Comment:
Here i tried copying in ONLY the buffer, which allows us to reconstruct the matrix inside the worker. But it ran quite slowly, even if we don't recreate the matrix on each run. I wasn't able to get it to run faster than non-workerized so I gave up!